### PR TITLE
Add __call__

### PR DESCRIPTION
**Pull request by EderSantana**
Add a `__call__` method to Layer and Sequential. We can do the following:
```python
X = K.placeholder(ndim=2)
Y = model(X)
```

One interesting application is where a model is used as an upper layer of another model. In that case, we don't have access to its `get_input`:
```python
encoder = Sequential()
decoder = Sequential()

model.add(encoder)
model.add(decoder)

X = K.placeholder(ndim=2)
Y = decoder(X)
```

This code temporally change the value of a specific element of the graph by `X` and call `get_output`. Regardless of the fact that the specific element is a `previous` or `input`. For example, the following code would crash:

```python
X = decoder.get_input()  # this is not a placeholder Tensor, it is a graph
Y = decoder.get_output()
F = function([X], Y)
```

I also updated some docs. Let me know what you think.

### CODE REVIEW

**Comment by EderSantana** [comment]
wait, I think my editor removed trailing white spaces that that are actually useful for markdown. Can you confirm that @fchollet ? Did I broke the docs?

File: keras/layers/core.py [hunk 1]
```diff
@@ -17,6 +17,19 @@
 import sys
 
 
+class Pass:
```
Comment (olegsinyavskiy) [inline keras/layers/core.py]: Hi, its a good practice to inherit from object always:
http://stackoverflow.com/questions/4015417/python-class-inherits-object

**Comment by EderSantana** [comment]
Thanks @olegsinyavskiy is this why I'm crashing under Python 3?

**Comment by olegsinyavskiy** [comment]
@EderSantana , I think you have outdated code that somehow doesn't work well with recent changes.
For example, the line you crash in "if isinstance(floatx, unicode):" is not present on master anymore (https://github.com/fchollet/keras/blob/master/keras/backend/common.py)

**Comment by EderSantana** [comment]
@olegsinyavskiy I see, I submitted that part too. It seems that it was changed a few minutes ago :D Let me rebase

**Comment by olegsinyavskiy** [comment]
@EderSantana , it might have been my first pull request, so I might be blamed for that:) For example, I changed the way yaml config is updated with 'sed'. Sorry! I'm going to track few pulls to see if my changes break other people code..

**Comment by fchollet** [comment]
> wait, I think my editor removed trailing white spaces that that are actually useful for markdown. Can you confirm that @fchollet ? Did I break the docs?

No. But whitespace changes in an unrelated PR should be avoided.

**Comment by EderSantana** [comment]
@olegsinyavskiy Could you help me to see what is going on with Python 3 tests?

File: keras/layers/core.py [hunk 2]
```diff
@@ -33,6 +46,25 @@ def __init__(self, **kwargs):
         if not hasattr(self, 'params'):
             self.params = []
 
+    def __call__(self, X):
```
Comment (olegsinyavskiy) [inline keras/layers/core.py]: @EderSantana , I'm new to keras, but its seems that this logic heavily relies on the order of checks that 'get_input' performs (e.g. check for previous, then for input etc.). Any change in get_input is going to break this logic. It would be nice to have some tests:)

**Comment by EderSantana** [comment]
If anyone wants to see `__call__` in practice, please check one of the last cells of this notebook: https://github.com/EderSantana/seya/blob/master/examples/Convolutional%20Variational%20AE.ipynb

**Comment by olegsinyavskiy** [comment]
@EderSantana , sure! I think master was in a broken state and https://github.com/fchollet/keras/commit/f295ecb302ea3298ab61dda892f323c7c7807d02 just fixed the master recently. So you better rebase again)

File: keras/layers/core.py (hunk 2 above)
Comment (EderSantana) [inline keras/layers/core.py]: I added one. An I just linked a working example. If you have some suggestions, please let me know. Also, if you have a better way and want to completely rewrite `__call__` I'm also ok with that.

Comment (olegsinyavskiy) [inline keras/layers/core.py]: @EderSantana , I don't see any new tests in this pull.. Maybe you forgot to commit it?
As for call overload, I'm not sure what it the policy on keras, but I would personally stay away on overloading 'call' in the core library - it proved to be really really hard to undo or change later in my practice. What ends up happening usually in big projects is that objects that are called with '()' are hard to track. Lets say get_input is changed in some non-obvious way so you have to make sure that 'call' is not going to break things. It is practically impossible to find all the calls: normal 'Find all: "functionX"' is not going to work. So there is always a small call somewhere that interacts in a weird way and breaks something else etc. Another case - if you for some reason decide to get rid of call or change arguments, you again will probably loose some of use cases because you just cant find it. This is a non-compiled language-specific issues since in python an invalid object call will happily just sit until someone actually calls it. So I would just rename it to 'apply_to_input' or add optional argument to 'get_output' or smth like that. Call overload is perfectly fine in the client code though - e.g. you need to send a callback to 'fit' function. However I'm not really aware of all the usecases and practices here, so other more experienced guys like @fchollet should know better:)

Comment (EderSantana) [inline keras/layers/core.py]: I used to use a method called `apply_model(model, X)` instead of `__call__` to do things like that, I just thought that `__call__` would make things cleaner and easier to use. Also `__call__` is the choice on TensorFlow, I thought this would also bring consistency.

**Comment by EderSantana** [comment]
@olegsinyavskiy what do you think of `test_call.py`??? let me know whenever you have time

@fchollet I added some tests, I've also been experimenting with this branch a lot and it seems to be working well. Is do you have any requests?

File: keras/layers/core.py [hunk 3]
```diff
@@ -17,6 +17,19 @@
 import sys
 
 
+class Pass(object):
```
Comment (fchollet) [inline keras/layers/core.py]: Not a fan of this. It is neither simple nor elegant.

**Comment by fchollet** [comment]
I think the API it introduces is potentially valuable. I see three issues.

- management of training/testing modes
- it may be confusing / complex to have different APIs to do similar things. We should have *one good way* to do each thing. 
- I am not a fan of the implementation, it seems to introduce quite a few hacks.

File: tests/keras/layers/test_call.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+# encoding: utf-8
+"""Test seya.layers.recurrent module"""
+
+from __future__ import print_function
+
+import unittest
```
Comment (olegsinyavskiy) [inline tests/keras/layers/test_call.py]: @EderSantana, thank you very much for the test! keras uses pytest so I personally don't see any point of using unittest classes. Unittest is too "java-ish", pytest is more modern framework.

**Comment by olegsinyavskiy** [comment]
To follow up on training modes comment of @fchollet:
Layers might have states. States might be updated during call (even if its not training mode), for that reason you have to put 'updates' argument into theano.function. This suggests that you probably need to return (Y, updates) tuple from the 'call' in the general case. I'm not sure how 'updates' concept is handled in tensorflow.. 

Example that comes to mind is RBM that might be implemented with theano.scan that is going to return non-trivial 'updates'. 

The good thing about keras in my opinion, is the hiding away symbolic manipulations from the user. This is the sharp contrast to 'blocks + fuel' framework that is a thin wrapper on top of the theano function. This leads to the really verbose (and IMHO ugly) 'blocks' api.

@EderSantana , looking at you ipynb, it seems that you might just use predict_classes, predict_probab without compiling the network yourself. Am I correct?

**Comment by elanmart** [comment]
@EderSantana shouldn't `__call__` take `train` argument?

FWIW when I needed something like this, I used lambda functions to avoid the need of creating the `Pass` class.

```python 
def apply(self, X, train=True):
        get_input = self.get_input
        self.get_input = lambda _: X
        output = self.get_output(train)
        self.get_input = get_input
        return output
```

**Comment by EderSantana** [comment]
@elanmart I was actually thinking about giving `__call__` a train argument. I'll do it. I'll also get how that works without Pass. Thanks for the feedback

About the tests, I made a mistake compiling the function using Theano function. I'll fix that too

**Comment by EderSantana** [comment]
@elanmart that was a killer suggestion man. Code and logic is much shorter now!!! Appreciate it!

**Comment by elanmart** [comment]
Cool! Glad I could help.

**Comment by EderSantana** [comment]
@fchollet I added tests and the `__call__` is much simpler and intuitive now. Also, it is working on both backends.

File: docs/sources/layers/core.md [hunk 5]
```diff
@@ -1,5 +1,7 @@
 ## Base class
 
+Note: Where ever we refer to a "Tensor" bear in mind that its the type depends on the backend you are using.
```
Comment (fchollet) [inline docs/sources/layers/core.md]: "its type"

File: docs/sources/layers/core.md [hunk 6]
```diff
@@ -7,6 +9,19 @@ keras.layers.core.Layer()
...
 ```python
+__call__
+```
+
+Apply Layer to an external input tensor.
+
+- __Return__: Tensor.
+
+- __Arguments__:
+    - __X__: Tensor. External input Tensor that will temporally replace
```
Comment (fchollet) [inline docs/sources/layers/core.md]: Meaning unclear.

File: keras/layers/core.py [hunk 7]
```diff
@@ -33,6 +33,15 @@ def __init__(self, **kwargs):
         if not hasattr(self, 'params'):
             self.params = []
 
+    def __call__(self, X, train=False):
+        # temporally substitute input
```
Comment (fchollet) [inline keras/layers/core.py]: "temporary"

File: tests/keras/layers/test_call.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+# encoding: utf-8
+"""Test seya.layers.recurrent module"""
```
Comment (fchollet) [inline tests/keras/layers/test_call.py]: Wrong package!

File: tests/keras/layers/test_call.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+import unittest
+import numpy as np
+
+from numpy.testing import assert_allclose
+
+from keras import backend as K
+from keras.layers.core import Dense
+from keras.models import Sequential
+floatX = K.common._FLOATX
```
Comment (fchollet) [inline tests/keras/layers/test_call.py]: `K.floatx()` is the preferred way to access this.

File: docs/sources/layers/core.md (hunk 6 above)
Comment (EderSantana) [inline docs/sources/layers/core.md]: How about "Apply layer transformation defined by its `get_output` method to an input `X`." Do you have any suggestion?

**Comment by fchollet** [comment]
Ok, looks good to me!

**Comment by fchollet** [comment]
This appears to have broken the build :( https://travis-ci.org/fchollet/keras/jobs/95116744

**Comment by fchollet** [comment]
It appears inconsequential. You may want to change the threshold of the assert allclose.

**Comment by EderSantana** [comment]
@fchollet Oh no I verified that on my local... Its problem with floatx I'm sure. The previous one was working ok. I believe we are having problems setting _FLOATX with a global. Also, it is hard corded at `backend` to float32. In travis we should assume float64.

Lets open a new issue to discuss that.

**Comment by fchollet** [comment]
> Also, it is hard corded at backend to float32. In travis we should assume float64.

I'd rather have Travis use float32. In theory we should test with *both*, though, as there are issues to be caught by switching from one to the other.

It's fine to have a higher threshold for assert allclose, I do that in many places.

**Comment by EderSantana** [comment]
Oh, I didn't know that happened. The test pass ok under float64. Let me increase the tol and make a new PR.
